qadence/backends/adjoint.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import annotations
+
+from typing import Any
+
+from pyqtorch.apply import apply_operator
+from pyqtorch.circuit import QuantumCircuit as PyQCircuit
+from pyqtorch.parametric import Parametric as PyQParametric
+from pyqtorch.primitive import Primitive as PyQPrimitive
+from pyqtorch.utils import overlap, param_dict
+from torch import Tensor, no_grad, tensor
+from torch.autograd import Function
+from torch.nn import Module
+
+from qadence.backends.pyqtorch.convert_ops import PyQHamiltonianEvolution, ScalePyQOperation
+from qadence.blocks.abstract import AbstractBlock
+
+
+class AdjointExpectation(Function):
+    @staticmethod
+    @no_grad()
+    def forward(
+        ctx: Any,
+        circuit: PyQCircuit,
+        observable: PyQCircuit,
+        state: Tensor,
+        param_names: list[str],
+        *param_values: Tensor,
+    ) -> Tensor:
+        for param in param_values:
+            param = param.detach()
+        ctx.circuit = circuit
+        ctx.observable = observable
+        ctx.param_names = param_names
+        values = param_dict(param_names, param_values)
+        ctx.out_state = circuit.run(state, values)
+        ctx.projected_state = observable.run(ctx.out_state, values)
+        ctx.save_for_backward(*param_values)
+        return overlap(ctx.out_state, ctx.projected_state)
+
+    @staticmethod
+    def backward(ctx: Any, grad_out: Tensor) -> tuple:
+        param_values = ctx.saved_tensors
+        values = param_dict(ctx.param_names, param_values)
+
+        def _apply_adjoint(ctx: Any, op: Module) -> list:
+            grads: list = []
+            if isinstance(op, (PyQHamiltonianEvolution)):
+                generator = op.block.generator
+                time_param = values[op.param_names[0]]
+
+                ctx.out_state = apply_operator(ctx.out_state, op.dagger(values), op.qubit_support)
```

Comment:
@dominikandreasseitz is this correct? You are daggering the generator and applying it to the state but `HamEvo` operation is the evolution of the generator. I might be missing something here. I would add a bit of comments about what is going on here because it is pretty difficult to decode.

Reply:
i mistakenly responded somewhere else, its the dagger of the evolved operator

Reply:
there is documentation in the pyqhamevo class